src/mesh.cpp

```diff
@@ -1502,6 +1534,597 @@ openmc_mesh_set_params(int32_t index, int n, const double* ll, const double* ur,
+                                          start.array(),
+                                          tris,
+                                          intersection_dists,
+                                          0,
+                                          track_len);
+  if (rval != moab::MB_SUCCESS) {
+    fatal_error("Failed to compute intersections on unstructured mesh: " + filename_);
+  }
+
+  // sort the hit triangles and intersections by distance
+  hits.clear();
+  for (int i = 0; i < tris.size(); i++) {
+    hits.push_back(std::pair<double, moab::EntityHandle>(intersection_dists[i], tris[i]));
+  }
+
+  // sorts by first component of std::pair by default
+  std::sort(hits.begin(), hits.end());
+}
+
+void
+UnstructuredMesh::bins_crossed(const Particle* p,
+                               std::vector<int>& bins,
+                               std::vector<double>& lengths) const {
+  moab::ErrorCode rval;
+  Position last_r{p->r_last_};
+  Position r{p->r()};
+  Position u{p->u()};
+  u /= u.norm();
+  moab::CartVect r0(last_r.x, last_r.y, last_r.z);
+  moab::CartVect r1(r.x, r.y, r.z);
+  moab::CartVect dir(u.x, u.y, u.z);
+
+  double track_len = (r1 - r0).length();
+
+  r0 += TINY_BIT*dir;
+  r1 -= TINY_BIT*dir;
+
+  UnstructuredMeshHits hits;
+  intersect_track(r0, dir, track_len, hits);
+
+  bins.clear();
+  lengths.clear();
+
+  // if there are no intersections the track may lie entirely
+  // within a single tet. If this is the case, apply entire
+  // score to that tet and return.
+  if (hits.size() == 0) {
+    moab::EntityHandle last_r_tet = get_tet(last_r + u * track_len * 0.5);
+    if (last_r_tet) {
+      bins.push_back(get_bin_from_ent_handle(last_r_tet));
+      lengths.push_back(1.0);
+    }
+    return;
+  }
+
+  // attempt to find the containing tet for the first track segment
+  moab::EntityHandle tet = get_tet(last_r + u * hits.front().first / 2.0);
+  double last_dist = 0.0;
+
+  // make sure first segment is inside a tet. If it is not, we may be starting
+  // outside of the mesh and our first intersection is an entry point into the
+  // mesh - update hits and find a containing tet accordingly
+  if (!tet) {
+    last_dist = hits.front().first;
+    hits.erase(hits.begin());
+    tet = get_tet(last_r + u * (last_dist + hits.front().first) / 2.0);
+  }
+
+  // if there are no other hits at this point, apply the score for whatever
+  // distance lies in this tet and return
+  if (hits.size() == 0 && tet) {
+    bins.push_back(get_bin_from_ent_handle(tet));
+    lengths.push_back((track_len - last_dist) / track_len);
+    return;
+  }
+
+  // score all remaining segments
+  for (const auto& hit : hits) {
+    // apply score in this tet if one was found
+    if (tet) {
+      bins.push_back(get_bin_from_ent_handle(tet));
+      lengths.push_back((hit.first - last_dist) / track_len);
+    } else {
+      // if in the loop, we should always find a tet
+      warning("No tet found for location between triangle hits");
```

Comment:
In a glancing intersection where a track intersects with the vertex of a tet (or directly on an edge of a face ) where we might not find a containing tet b/c it lies right on the boundary of that element. The presumption here is that this is ok because the track length in that element would be near-zero, but I thought if there were something systematically wrong it would be best to let the user know if this is happening frequently.

As the comment above this line indicates, I originally thought it should be an error.